### DISCUSSION TITLE
.NET 7

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#2465] [General]
The release of .NET 7 is less than a month away - it will be released on Nov 8 as part of .NET Conf 2022. That being said, I would like to hear some feedback from the community in regards to how the .NET release schedule should be factored into Oqtane's roadmap.

For those who do not understand the current .NET support model... 

- .NET 6 (which Oqtane is currently on) is an LTS release which stands for Long Term Support and means that the release is fully supported for 3 years (which means it is valid until November 2024)
-  .NET 7 will be a Current release which means it will only be supported for 18 months (valid until May 2024)
- .NET 8 will be an LTS release which is expected to be released November 2023

In general, most enterprise organizations prefer LTS releases as they provide the most long term stability. The adoption of .NET 6 has been massive over the past year due to the fact that organizations which had been running on the last LTS release (3.1) had largely ignored .NET 5 in favor of waiting for the next LTS release (.NET 6). This being said, there are always exceptions - especially in cases where a Current release has specific features or capabilities which an enterprise needs to meet a business objective. There are certainly many improvements in .NET 7... many of which Microsoft has been promoting in recent months which will be showcased at .NET Conf 2022. However, the real question to be considered is what makes the most sense for Oqtane.

Traditionally, Oqtane has released a new major version at the same time as a new release of .NET... Oqtane 1.0 (.NET Core 3.1), Oqtane 2.0 (.NET 5), Oqtane 3.0 (.NET 6). This made a lot of sense as Oqtane is based on Blazor and Blazor was a new technology that was rapidly evolving during this period. In .NET 6 it could be argued that Blazor reached a level of maturity and stability where it could be considered a fully matured enterprise product offering. Microsoft even added Blazor Hybrid support as a late addition to .NET 6 to ensure enterprises could take advantage of the desktop and mobile hosting models under an LTS banner. At this stage there is nothing specific in .NET 7 which Oqtane needs to be successful... .NET 6 has all of the essential features. 

If Oqtane were to upgrade to .NET 7 at the time of release, it means that run-time environments would also need to be upgraded to .NET 7 in order to be able to run the framework. Third party modules and themes would not need to upgrade to .NET 7 immediately - they should continue to function just fine as the Oqtane APIs would remain backward compatible.  But once the Oqtane Dev branch is migrated to .NET 7, enterprises who want to remain on .NET 6 (for the LTS benefit) would only be able to utilize the last official Oqtane 3 release (version 3.2.1 as of now). Of course it is technically possible to maintain multiple branches in parallel (ie. a .NET 7 branch and a .NET 6 branch) however based on past experience this requires a lot of additional work and is not feasible for a project like Oqtane which currently has a small number of contributors. 

So how do we ensure the maximum number of people are able to use the framework while still allowing it to stay near the cutting edge of technology? Perhaps the best plan would be to wait for 6 months to determine the adoption of .NET 7 and make a judgement call at that time. The expectation is that enterprises will once again flock to .NET 8 in November of 2023 (because it will be an LTS release).

Thoughts?

**Comment by markdav-is** [comment]
If there are compelling improvements to MAUI Blazor hybrid capabilities in 7 then I'd say keep the lock-step with dev moving to Oqtane 4 on .Net 7.  Another thing to consider is if Azure App Services will be supporting .Net 7 on day-one.  If not, then that's a good reason to stay on 6 until 7 is supported in Azure.  Many of the blazor improvements touted in 7 are WASM and I've never tried to deploy Oqtane WASM style (but I'd like to try)

**Comment by sbwalker** [comment]
The main Blazor improvements in .NET 7 are below (not sure if all of them made the final cut):

![image](https://user-images.githubusercontent.com/4840590/196963470-bd195618-cdc4-49f2-bb75-4c51375a27e2.png)

Like I said, there is nothing specific on this list which Oqtane has been waiting for.

**Reply by markdav-is** [reply]
thanks.  I don't see anything overly interesting besides custom elements that would apply to Oqtane.

**Comment by JanOlsmar** [comment]
I prefer to use and look for the latest. Why stay behind :)

**Comment by ADefWebserver** [comment]
The advantage of waiting for an LTS (Long Term Support) release, and making that an Oqtane requirement, is that a lot of people like to deploy an Oqtane site and then not have to upgrade it for years.

**Comment by sbwalker** [comment]
A lot of good feedback in this thread. .NET 7 launches today and if you have been following the PRs you will notice that the framework has not yet been upgraded to .NET 7. I expect the effort to upgrade the framework to .NET 7 to be trivial... however it is the ripple effect of this upgrade which has the potential to cause challenges as there are many external modules, themes, environments which are all dependent upon the framework. So rather than upgrading for the sake of upgrading - which is not a strategy - in the near term we will wait and see if some real use cases emerge which necessitate an upgrade to .NET 7.

**Comment by iJungleboy** [comment]
I believe there is a case for both models, but let's have a slightly closer look...

## Current Developer Profile

To be honest most Oqtane devs are currently "bleeding-edge" users who know what this means. They are _not_ .net Framework developers who run servers that should never be upgraded, but instead are forced to install the latest .net 6 that's only a few months old. 

For example, two companies we're consulting are both developing the next version of they long established product and once it's done, it should be using the latest/greatest so it can again weather the storms for many years to come. They want the latest, because they have been stuck on an old stack for more than 10 years and want to get out of this freezes. 

## Option 1: Remain on .net 6 means Robust Technology Slow-Down

This is of course a very robust choice, but it also means that it stops innovation and benefits from platform improvements in it's tracks. 

...for 1-2 years. Because that's how I understand the tech cycle will be. .net 6 is LTS so it won't get many improvements any more and feels a lot like .net 4.8 which is missing so much incl. newer C# language features. 

## Option 2: Go .net 7 means Innovation, latest features, latest C# language etc.

Going for .net 7 means it will attract more developers. Period. Because they want to learn the ropes, because the want the latest etc. Or because they were stuck on .net 4.8 for years and now want the latest - and may feel that .net 6 is sooo 2022. 

## My Recommendation: Go strong

Oqtane is still so young, and the audience is very latest-tech. 
I would feel that we should give them the latest. 

Possibly create an Oqtane 3.3 LTS which would be the last Oqtane 3 so they don't have to hurry - also do some minimal security support, but otherwise move forward. 

There will come a day where we should move more slowly, but I'm convinced that day has not arrived at all.

**Reply by ADefWebserver** [reply]
Ok I am now convinced and in the "Go strong" camp. The fact that Azure is already on .Net 7 means that most end-users deploying Oqtane would not have 'problems' deploying Oqtane.

**Comment by sbwalker** [comment]
@iJungleboy it might not have been clear in my original post but .NET 8 (LTS) will be released in Nov 2023 - and since .NET 7 was just released this week it means that .NET 7 will only be valid for a maximum of 1 year (nobody will want to use it beyond Nov 2023 as it will not be supported - whereas .NET 6 will continue to be supported until Nov 2024). 

My perspective may be a bit biased on this topic, as I am used to working with very large enterprise clients... and in general they will not migrate to .NET 7 because of the short support window. And since they tend to have a large degree of influence over the rest of the tech industry, the rest of the industry tends to cater to their preferences. In addition, I have yet to encounter any feature available in.NET 7 which is a "must have" in Oqtane. Like most mature technology, later stage enhancements tend to cater more to edge cases than mainstream scenarios... and this is what I have noticed is the case with .NET Core lately as well (since it is V6 already - as V4 was skipped).

All that being said, I think the best approach is to "watch and wait" to see the actual adoption of .NET 7 in the coming months - and then make decisions based on facts.

**Reply by ADefWebserver** [reply]
@sbwalker You have explained your reasoning well. The performance gains, particularly with EF Core 7, seems important to consider. The question of course is: 'are the performance gains significant enough to be worth it?'. I guess it doesn't matter how much the benchmark gains are, it is the Oqtnae performance gains that matter.

**Reply by sbwalker** [reply]
@ADefWebserver I have been following the EF Core performance enhancements... and they look interesting - the one thing I have not been able to figure out is if they are only applicable to SQL Server.

**Comment by iJungleboy** [comment]
Without knowing the details, I'm pretty sure AOT and pre-render is also much better in .net 7. 

This could be very important for SEO, because the current `display: none` in the page is a big issue for SEO, and it's only there because of the flash of content on-load, which is in turn because pre-render isn't really good yet in .net 6.

**Reply by iJungleboy** [reply]
I also noticed that the docs now say 

> The rendered UI also matches the prerendered UI, so no flicker occurs in the browser

https://learn.microsoft.com/en-us/aspnet/core/blazor/components/prerendering-and-integration?view=aspnetcore-7.0&pivots=server

that would be really, really important.

**Reply by sbwalker** [reply]
Unfortunately in order to follow the guidance it requires a non-trivial code change in every component - so it would not just affect all of the framework admin module components but also the components in every third party module. This does not seem like a feasible solution.

**Comment by iJungleboy** [comment]
Just a heads up: NopCommerce is now also moving to .net 7 in v4.6

Just as a comparison, because it is a huge platform with wide adoption, and is also pursuing a go-strong strategy: https://www.nopcommerce.com/en/boards/topic/95663/nopcommerce-4600-beta-released-please-share-your-impressions

**Comment by Tony20221** [comment]
Every newer version of .NET has better performance improvement without any changes in code in the application. I say don't use any C# 11 and Blazor features that require .NET 7 until .NET 8 is out and after a few months (6 as you suggested) of 8's release start using .NET 8 features. So that people who only use LTS releases can stay on .NET 6 these days until .NET 8 is out.
Isn't targeting .NET 6 or 7 is just changing the target framework and recompiling?

**Comment by leigh-pointer** [comment]
Interesting https://youtu.be/QVkxusemLoo?t=3570 Upgrading 6-7-8 or 6-8

**Reply by sbwalker** [reply]
Thank you for sharing!

**Comment by sbwalker** [comment]
I have been monitoring Oqtane's dependencies in terms of their support for .NET 7 and the only dependency which does not have an official  .NET 7 release yet is MySql.EntityFrameworkCore) (Preview5). so I am thinking that a .NET 7 release of Oqtane is feasible in Q1 2023.

**Reply by JanOlsmar** [reply]
Sounds like a great plan.

**Comment by Sharafudeen** [comment]
There are lots of benefits to migrate dotnet 7 but in real world scenario many organizations they might not migrate enterprise applications dependencies which is NOT LTS version. My humble suggestion is to upgrade LTS version of framework. If people really wants then better to provide an option in Oqtane to upgrade NOT LTS version. This will be more flexible for Oqtane users :)
